### PR TITLE
fix: doctor for 0.72 Ruby changes

### PR DESCRIPTION
**Pull request by blakef**
Summary:
---------

React Native 0.72.0 requires support for a range of Ruby version.

**Fixes:**
- We were using the version of `gem` instead of `ruby` to validate against.  Moved to `ruby`.
- Supports `Gemfiles` and `.ruby-version` (as a fallback) and finally the version hardcoded into doctor.
- Supports more error cases

**Other:**
- Added unit tests
- Added some inline documentation about why React Native has gone back to supporting system Ruby (but includes a range over versions to support).

# Examples:
## Version doesn't match Gemfile requirements
CleanShot 2023-03-03 at 17 39 11

## No Gemfile
CleanShot 2023-03-03 at 17 22 52

## No Gemfile for .ruby-version file
CleanShot 2023-03-03 at 17 22 24

Test Plan:
----------

I ran this against a local project, adjusting variations of the `Gemfile` and `.ruby-version`.

### CODE REVIEW

File: packages/cli-doctor/src/tools/healthchecks/ruby.ts [hunk 1]
```diff
@@ -1,19 +1,211 @@
...
+  NO_RUBY: 'No Ruby',
+  BUNDLE_INVALID_RUBY: 'Bundle invalid Ruby',
+  UNKNOWN: 'Unknown',
+} as const;
+
+// The Change:
+// -----------
+//
+// React Native 0.72 primarily defines the compatible version of Ruby in the
+// project's Gemfile [1]. It does this because it allows for ranges insteado of
```
Comment (cipolleschi) [inline packages/cli-doctor/src/tools/healthchecks/ruby.ts:26]:
```suggestion
// project's Gemfile [1]. It does this because it allows for ranges instead of
```

File: packages/cli-doctor/src/tools/healthchecks/ruby.ts [hunk 2]
```diff
@@ -1,19 +1,211 @@
...
+// having to pin to a version of Ruby.
```
Comment (cipolleschi) [inline packages/cli-doctor/src/tools/healthchecks/ruby.ts:27]:
```suggestion
// pinning to a version of Ruby.
```

File: packages/cli-doctor/src/tools/healthchecks/ruby.ts [hunk 3]
```diff
@@ -1,19 +1,211 @@
...
+//
+// In previous versions the .ruby-version file defined the compatible version,
+// and it was derived in the Gemfile [2]:
+//
+// > ruby File.read(File.join(__dir__, '.ruby-version')).strip
+//
+// Why all of the changes with Ruby?
+// ---------------------------------
+//
+// React Native has had to weigh up a couple of valid concerns:
```
Comment (cipolleschi) [inline packages/cli-doctor/src/tools/healthchecks/ruby.ts:37]:
```suggestion
// React Native has had to weigh up a couple of concerns:
```

File: packages/cli-doctor/src/tools/healthchecks/ruby.ts [hunk 4]
```diff
@@ -1,19 +1,211 @@
...
+//
+// - Cocoapods: we don't control the minimum supported version, although that
+//   was defined almost a decade ago [3]. Practically system Ruby on macOS works
+//   for our users.
+//
+// - Apple may drop support for scripting language runtimes in future version of
+//   macOS [4]. Ruby 2.7 is effectively EOL, which means many supporting tools and
+//   developer environments _may_ not support it going forward, and 3.0 is becoming
+//   the default in, for example, places like our CI. Some users may be unable to
```
Comment (cipolleschi) [inline packages/cli-doctor/src/tools/healthchecks/ruby.ts:46]:
```suggestion
//   the default in, for example, places like our CI. Some users may be unable to
```

That's not entirely true. CircleCI still supports Ruby 2.7.7... we don't know what will happens in April though.

File: packages/cli-doctor/src/tools/healthchecks/ruby.ts [hunk 5]
```diff
@@ -1,19 +1,211 @@
...
+//   install Ruby 2.7 on their devices as a matter of policy.
+//
+// - Our Codegen is extensively built in Ruby 2.7.
```
Comment (cipolleschi) [inline packages/cli-doctor/src/tools/healthchecks/ruby.ts:49]: I won't callout Codegen specifically. I'd say that we have some build scripts for iOs that are written in ruby and we used Ruby 2.7 to develop them.

**Review by cipolleschi — COMMENTED** [review]
Hi there, I left some comments mostly on the documentation part. I'll leave the technical review to someone who is more TS-savvy then me! :D

File: packages/cli-doctor/src/tools/envinfo.ts [hunk 6]
```diff
@@ -33,8 +33,8 @@ async function getEnvironmentInfo(
       System: ['OS', 'CPU', 'Memory', 'Shell'],
       Binaries: ['Node', 'Yarn', 'npm', 'Watchman'],
       IDEs: ['Xcode', 'Android Studio', 'Visual Studio'],
-      Managers: ['CocoaPods', 'RubyGems'],
-      Languages: ['Java'],
+      Managers: ['CocoaPods'],
+      Languages: ['Java', 'Ruby'],
```
Comment (arushikesarwani94) [inline packages/cli-doctor/src/tools/envinfo.ts:37]: This change was also discussed in : https://github.com/react-native-community/cli/issues/1825

Whether we need `RubyGems` in Managers or `Ruby` in Languages

Comment (blakef) [reply packages/cli-doctor/src/tools/envinfo.ts:37]: Leaving this here to follow up: https://github.com/react-native-community/cli/pull/1818#discussion_r1097095344

There was some discussion about this in that thread. I think this tackles some of the issues raised by focusing on the Gemfile. Interestingly I'm not sure it's useful to consider the .ruby-version at all, even as a backup. 

## Tasks
- [ ] Can we build without a Gemfile (I don't think so)
- [ ] We could try fetch the actual requirement from the tagged releases, but I see lots of avenues for failure here (it's just tricky finding a way to grab the correct version for the matching version of RN - 2 network requests there at a minimum).

**Comment by adamTrz** [comment]
Hi @blakef thanks for PR 💯 
I'll try to review it shortly, one thing that comes in mind tho right now - IIRC `ruby-version` file will be removed in 0.72 via [this PR](https://github.com/facebook/react-native/pull/36281) so don't think we can care about it if doctor change is targeting 72? WDYT?

**Comment by blakef** [comment]
> Hi @blakef thanks for PR 💯 I'll try to review it shortly, one thing that comes in mind tho right now - IIRC `ruby-version` file will be removed in 0.72 via [this PR](https://github.com/facebook/react-native/pull/36281) so don't think we can care about it if doctor change is targeting 72? WDYT?

I touched on this a little when responding to @arushikesarwani94's comment.  Originally I was concerned about people using the doctor on older projects, but now thinking about it they still have the `Gemfile` (which typically would have just loaded the value from the `.ruby-version`).

**TL;DR: agree, we should pull out all references to `.ruby-version`.**

File: packages/cli-doctor/src/tools/versionRanges.ts [hunk 7]
```diff
@@ -3,7 +3,7 @@ export default {
   NODE_JS: '>= 16',
   YARN: '>= 1.10.x',
   NPM: '>= 4.x',
-  RUBY: '>= 2.7.6',
+  RUBY: '>= 2.7.10',
```
Comment (cortinico) [inline packages/cli-doctor/src/tools/versionRanges.ts:6]:
```suggestion
  RUBY: '>= 2.6.10',
```

This should align with: https://github.com/facebook/react-native/blob/main/template/Gemfile

cc @cipolleschi

**Review by cortinico — CHANGES_REQUESTED** [review]

File: packages/cli-doctor/src/tools/envinfo.ts (hunk 6 above)
Comment (cipolleschi) [reply packages/cli-doctor/src/tools/envinfo.ts:37]: > Can we build without a Gemfile (I don't think so)

We can but we would not have any guarantee on the version of Cocoapod used by the users. So, if they have an old version of Cocoapods installed in the current Ruby, they will be using that which may be not compatible with React Native.

So, I highly discourage that we don't use Gemfile.

**Review by cortinico — APPROVED** [review]
Accepting to unblock you @arushikesarwani94 but there are still a couple of open points from @cipolleschi that needs to be addressed

**Comment by blakef** [comment]
I've reverted the update to the test environment, since the current release (0.71.3) pins Ruby to 2.7.6.  Once 0.72.0 is released, that should be changed to 2.6.10.

**Review by adamTrz — APPROVED** [review]
